Fix compatibility of gengo with Windows

ricky26 opened the pull request:
Swaps some usages of path and filepath and replaces some usages of PathSeperator with "/" in order to make consistent usage of slash-paths and filepaths in gengo. Also fixes up a couple of unit tests which fail on Windows otherwise.

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/gengo/101)
<!-- Reviewable:end -->

ricky26 commented:
@k8s-ci-robot I think I've signed that now!

sttts commented on generator/import_tracker.go:
```diff
@@ -37,7 +37,12 @@ func NewImportTracker(typesToAdd ...*types.Type) namer.ImportTracker {
+		glog.Warningf("Warning: backslash used in import path '%v', this is unsupported.\n", path)
+	}
+
+	dirs := strings.Split(path, "/")
```
a goseparator const would make this much more explicit.

sttts commented on generator/import_tracker.go:
```diff
@@ -37,7 +37,12 @@ func NewImportTracker(typesToAdd ...*types.Type) namer.ImportTracker {
 	path := t.Package
-	dirs := strings.Split(path, string(filepath.Separator))
+
+	if strings.ContainsRune(path, '\\') {
```
and put a link to that discussion above the const I suggested below. Then we only need it once.

sttts commented on namer/namer.go:
```diff
@@ -23,6 +23,17 @@ import (
 
+const (
+	// GoSeperator is used to split go import paths.
+	// Forward slash is used intead of filepath.Seperator because it is the
```
instead

sttts commented on namer/namer.go:
```diff
@@ -187,7 +198,7 @@ var (
 // filters out unwanted directory names and sanitizes remaining names.
 func (ns *NameStrategy) filterDirs(path string) []string {
-	allDirs := strings.Split(path, string(filepath.Separator))
+	allDirs := strings.Split(path, goseperator)
```
GoSeperator, no?

sttts commented on parser/parse.go:
```diff
@@ -488,7 +489,9 @@ func (b *Builder) findTypesIn(pkgPath importPathString, u *types.Universe) error
 	for _, f := range b.parsed[pkgPath] {
-		if strings.HasSuffix(f.name, "/doc.go") {
+		_, fileName := filepath.Split(f.name)
+
```
nit: no empty line.

Even better: `if _, fileName := filepath.Split(f.name); fileName == "doc.go" {`

sttts commented on generator/import_tracker.go:
```diff
@@ -17,9 +17,9 @@ limitations under the License.
 	"strings"
 
+	"github.com/golang/glog"
 	"k8s.io/gengo/namer"
```
nit: new line above

sttts commented:
To avoid that windows support breaks in the future, is there a travis like free CI for Windows?

ricky26 commented:
I've heard of AppVeyor being used but I've not used it myself.

ricky26 commented:
Review status: 0 of 5 files reviewed at latest revision, 6 unresolved discussions.

---

*[generator/import_tracker.go, line 41 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/gengo/101#-L5rf57UwECrQ6N1KqYW-r1-41:-L6MWtv4iIRPAYB4ClOk:b-dychkr) ([raw file](https://github.com/kubernetes/gengo/blob/934f01b29f2940ef41d2b5f76b9d72f10b5e7753/generator/import_tracker.go#L41)):*
<details><summary><i>Previously, sttts (Dr. Stefan Schimanski) wrote…</i></summary><blockquote>

and put a link to that discussion above the const I suggested below. Then we only need it once.
</blockquote></details>

I've stuck a comment over the specific backslash check and added the rest over a GoSeperator constant.

---

*[generator/import_tracker.go, line 45 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/gengo/101#-L5rf57UwECrQ6N1KqYW-r1-45:-L6MX2BR3EacqFEXgrIA:b-8q9xtq) ([raw file](https://github.com/kubernetes/gengo/blob/934f01b29f2940ef41d2b5f76b9d72f10b5e7753/generator/import_tracker.go#L45)):*
<details><summary><i>Previously, sttts (Dr. Stefan Schimanski) wrote…</i></summary><blockquote>

a goseparator const would make this much more explicit.
</blockquote></details>

Done, I've dropped this in the namer package, since that seems to be used by import_tracker where the other usage is.

---

*[generator/import_tracker.go, line 23 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/gengo/101#-L5rf57UwECrQ6N1KqYW-r2-23:-L6MXJnbjm0KyVhdjd8q:b-896fix) ([raw file](https://github.com/kubernetes/gengo/blob/580f9b9ff128a5344c2136f84fe8b4397208f4c5/generator/import_tracker.go#L23)):*
<details><summary><i>Previously, sttts (Dr. Stefan Schimanski) wrote…</i></summary><blockquote>

nit: new line above
</blockquote></details>

Done.

---

*[namer/namer.go, line 28 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/gengo/101#-L5rf57UwECrQ6N1KqYX-r2-28:-L6MX7e5woJSpzP6P-L7:b-896fix) ([raw file](https://github.com/kubernetes/gengo/blob/580f9b9ff128a5344c2136f84fe8b4397208f4c5/namer/namer.go#L28)):*
<details><summary><i>Previously, sttts (Dr. Stefan Schimanski) wrote…</i></summary><blockquote>

instead
</blockquote></details>

Done.

---

*[namer/namer.go, line 201 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/gengo/101#-L5rf57UwECrQ6N1KqYX-r2-201:-L6MX8Lonwl8NBnty6f4:b-896fix) ([raw file](https://github.com/kubernetes/gengo/blob/580f9b9ff128a5344c2136f84fe8b4397208f4c5/namer/namer.go#L201)):*
<details><summary><i>Previously, sttts (Dr. Stefan Schimanski) wrote…</i></summary><blockquote>

GoSeperator, no?
</blockquote></details>

Done.

---

*[parser/parse.go, line 493 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/gengo/101#-L5rf57UwECrQ6N1KqYY-r2-493:-L6MXBOMR9jkM_sm6n7X:b-896fix) ([raw file](https://github.com/kubernetes/gengo/blob/580f9b9ff128a5344c2136f84fe8b4397208f4c5/parser/parse.go#L493)):*
<details><summary><i>Previously, sttts (Dr. Stefan Schimanski) wrote…</i></summary><blockquote>

nit: no empty line.

Even better: `if _, fileName := filepath.Split(f.name); fileName == "doc.go" {`
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/gengo/101)*
<!-- Sent from Reviewable.io -->

sttts commented:
@ricky26 can you try to get AppVeyor working with this repo? I would feel better with CI in place.

ricky26 commented:
@sttts I've put together an appveyor.yml. Someone with webhook access to the project will need to sign up for an AppVeyor account.

You can check the output [here](https://ci.appveyor.com/project/ricky26/gengo/history). As far as I can tell it's now successfully running all the steps that Travis is.

sttts commented on appveyor.yml:
```diff
@@ -0,0 +1,34 @@
+
+environment:
+  GOPATH: c:\gopath
+  GOVERSION: 1.8
```
can we have 1.10 here?

sttts commented:
Looks great. Can you squash the changes? Then it's ready to go.

@thockin do you have permissions to setup the AppVeyor hook? That service is free for OSS.

thockin commented:
I am sure I have permissions.  Where are instructions.

ricky26 commented:
@thockin If you sign up on appveyor.com and follow the new project wizard for github it adds the hooks for you.

There are a few settings in the project once you've set it up such as how it numbers its builds but I don't think we'll need to change any of them if we're just using it for testing.

ricky26 commented:
@thockin have you had any opportunity to look at this?

sttts commented:
How can we make progress here?

ricky26 commented:
@sttts I'm not sure - the changes are working from my point of view (including the appveyor script).

ricky26 commented:
@lavalamp done!

lavalamp commented:
/lgtm
/approve
